### ISSUE TITLE
Possibility of a for_each_init() like function that returns a Vec containing the init variables.

### ISSUE THREAD

**Issue by MarcusTL12** [rayon-rs/rayon#955]
I have looked through the docs, and issues, but cannot seem to find a function in rayon for this use case. I am constructing a matrix in parallel where all threads needs random access to the entire matrix, but if each thread has their own matrix, then I can just add together the resulting matrices. Currently I am using the [thread_local](https://crates.io/crates/thread_local) crate, which works fine, but I feel that it would be nice to have something like:
```rust
let thread_local_data: Vec<_> = some_par_iter.for_each_init(
    || Array2::zeros(m, n),
    |mat, iter_data| // mess with mat in some way
)
.collect();

// sum up the matrices in thread_local_data
```

From looking at #718 and #742 I realize that the for_each_init function is not just simply calling init once for each thread, so this might also not be possible, but I think it would be really nice to have a feature like this. For now though I will keep using the thread_local crate, as that does what I need more explicitly, even though having to deal with RefCell is not as pretty.

**Comment by cuviper** [comment]
I think `.fold(...).collect()` will do what you want, or instead of `collect` you could use `reduce` or `reduce_with` to combine them on the fly.

**Comment by MarcusTL12** [comment]
Ah, I agree. `.fold(...)` and `.fold_with(...)` solves my problem.
